arrow-fx-stm/src/main/kotlin/arrow/fx/stm/TQueue.kt

```diff
@@ -7,17 +7,214 @@ fun <A> STM.newTQueue(): TQueue<A> = TQueue(newTVar(PList.Nil), newTVar(PList.Ni
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.stm.TQueue
+ * import arrow.fx.stm.atomically
+ *
+ * suspend fun main() {
+ *   //sampleStart
+ *   val tq = TQueue.new<Int>()
+ *   val result = atomically {
+ *     tq.write(2)
+ *
+ *     tq.peek()
+ *   }
+ *   //sampleEnd
+ *   println("Result $result")
+ *   println("Items in queue ${atomically { tq.flush() }}")
+ * }
+ * ```
+ *
+ * As with [STM.read] [STM.peek] will retry should the queue be empty. The alternative [STM.tryPeek] is there to avoid that:
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.stm.TQueue
+ * import arrow.fx.stm.atomically
+ *
+ * suspend fun main() {
+ *   //sampleStart
+ *   val tq = TQueue.new<Int>()
+ *   val result = atomically {
+ *     tq.tryPeek()
+ *   }
+ *   //sampleEnd
+ *   println("Result $result")
+ *   println("Items in queue ${atomically { tq.flush() }}")
+ * }
+ * ```
+ *
+ * It is also possible to read the entire list in one go using [STM.flush]:
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.stm.TQueue
+ * import arrow.fx.stm.atomically
+ *
+ * suspend fun main() {
+ *   //sampleStart
+ *   val tq = TQueue.new<Int>()
+ *   val result = atomically {
+ *     tq.write(2)
+ *     tq.write(4)
+ *
+ *     tq.flush()
+ *   }
+ *   //sampleEnd
+ *   println("Result $result")
+ *   println("Items in queue ${atomically { tq.flush() }}")
+ * }
+ * ```
+ *
+ * ## Checking a queues size
+ *
+ * Checking if a queue is empty can be done by using either [STM.isEmpty] or [STM.isNotEmpty]:
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.stm.TQueue
+ * import arrow.fx.stm.atomically
+ *
+ * suspend fun main() {
+ *   //sampleStart
+ *   val tq = TQueue.new<Int>()
+ *   val result = atomically {
+ *     tq.isEmpty()
+ *   }
+ *   //sampleEnd
+ *   println("Result $result")
+ * }
+ * ```
+ *
+ * Retrieving the actual size of a list can be done using [STM.size]:
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.stm.TQueue
+ * import arrow.fx.stm.atomically
+ *
+ * suspend fun main() {
+ *   //sampleStart
+ *   val tq = TQueue.new<Int>()
+ *   val result = atomically {
+ *     tq.size()
+ *   }
+ *   //sampleEnd
+ *   println("Result $result")
+ * }
+ * ```
+ *
+ * > All three of these methods have to access both the write and read end of a [TQueue] and thus can increase contention. Use them sparingly!
+ *
+ * ## Filtering a [TQueue]
+ *
+ * It is also possible to (destructively) filter a [TQueue] using [STM.filter] and [STM.filterNot]:
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.stm.TQueue
+ * import arrow.fx.stm.atomically
  *
- * [TQueue] is implemented using two [TVar]'s. One for reading and one for writing.
- * This effectively means that writing and reading accesses two different variables and thus it never blocks each other.
+ * suspend fun main() {
+ *   //sampleStart
+ *   val tq = TQueue.new<Int>()
+ *   atomically {
+ *     tq.write(0)
+ *     tq.filter { it != 0 }
```

Comment:
filter as a name is confusing if it does not return a new queue as users may be used ro. Is there a better name for this op?

Reply:
I honestly have no idea, it really does filter, so dunno. Actually I really dislike the filter as an api because it a) increases contention (accesses both TVars) and b) its not pure and has a weird/different api as you mentioned.

Maybe it makes sense to just :fire: it.

Reply:
Maybe `applyFilter` to highlight that it directly applies to the `TQueue`?

Reply:
If you dislike the function, I am fine with removing it. It can always be re-added later, removing is harder 😁
I agree the name is a bit confusing.
Perhaps something along the line of `remove`/`removeAll`? That's a function found in the Kotlin Standard library but only available for mutable collections.

https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/remove.html
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/remove-all.html

Reply:
I think keeping it is fine, `removeAll` with a predicate is a really good name. It still has the problem of contention but it may have some actual use, so long as its used infrequently its perfectly fine.